src/db.c

```diff
@@ -1340,6 +1294,51 @@ int dbSwapDatabases(int id1, int id2) {
     return C_OK;
 }
 
+/* Logically, this discards (flushes) the old main database, and apply the newly loaded
+ * database (temp) as the main (active) database, the actual freeing of old database
+ * (which will now be placed in the temp one) is done later. */
+void swapMainDbWithTempDb(tempDb *tempDb) {
+    for (int i=0; i<server.dbnum; i++) {
+        redisDb aux = server.db[i];
+        redisDb *activedb = &server.db[i], *newdb = &(tempDb->dbarray[i]);
+
+        /* Swap hash tables. Note that we don't swap blocking_keys,
+        * ready_keys and watched_keys, since clients 
+        * remain in the same DB they were. */
+        activedb->dict = newdb->dict;
+        activedb->expires = newdb->expires;
+        activedb->avg_ttl = newdb->avg_ttl;
+        activedb->expires_cursor = newdb->expires_cursor;
+
+        newdb->dict = aux.dict;
+        newdb->expires = aux.expires;
+        newdb->avg_ttl = aux.avg_ttl;
+        newdb->expires_cursor = aux.expires_cursor;
+
+        /* Now we need to handle clients blocked on lists: as an effect
+        * of swapping the two DBs, a client that was waiting for list
+        * X in a given DB, may now actually be unblocked if X happens
+        * to exist in the new version of the DB, after the swap.
+        *
+        * However normally we only do this check for efficiency reasons
+        * in dbAdd() when a list is created. So here we need to rescan
+        * the list of clients blocked on lists and signal lists as ready
+        * if needed.
+        *
+        * Also the swapdb should make transaction fail if there is any
+        * client watching keys */
+        scanDatabaseForReadyLists(activedb);
+        touchAllWatchedKeysInDb(activedb, newdb);
+    }
+
+    /* Restore slots to keys map if enable cluster. */
+    if (server.cluster_enabled) {
+        memcpy(server.cluster->slots_to_keys, tempDb->slots_to_keys, sizeof(server.cluster->slots_to_keys));
+    }
+
+    server.dirty++;
```

Comment:
I have raised this question to @zuiderkwast here: https://github.com/redis/redis/pull/9323#issuecomment-918484591

Reply:
It's right, but it's better if the implentation of slot-to-key can be hidden in cluster.c. slotToKeyCopyToBackup and slotToKeyRestoreBackup did this.

I have an idea about this. I'll comment more tomorrow.

Reply:
@eduardobr i was asking about the `dirty++`

Reply:
Oh, yes. I saw this in many places where db is touched and felt uncomfortable in not having it here as well.

Reply:
i was thinking for a moment that this should actually be incremented by the number of keys in the db, and that this should certainly be done in any diskless replication rather than just on swapdb mode.

i.e. in disk-based replication, the rdb we got from the master serves for persistence too, so server.dirty can be set to 0, but in diskless, it should be incremented for each key we load.
(somewhat related to https://github.com/redis/redis/pull/6217)

however, i see that on disk-based we don't reset it to 0, so i'm not certain if we should touch it in diskless at all.
@yossigo @soloestoy @madolson please let me know if there's anything i'm missing.

Reply:
@oranagra The save policy refers to "write operations", not sure if loading an entire rdb should be considered as N operations depending on number of keys or just a single operation. I lean towards leaving it as is.

Reply:
ok, by "as is" i assume you mean to increment the counter only once.
i.e. before this PR it wasn't incremented at all.
and if we do that, we should increment it also in any of the other diskless loading modes (not just swapdb / async).
@eduardobr please make it happen, and also mention this "bugfix" in the top comment of the PR (will be used for release notes and commit comment when squash-merging)

Reply:
Done and PR description updated

Reply:
I checked the codes about `server.dirty`, sadly its usages and definitions are ambiguous(I can't figure it out it means how many keys changed or event happened), and expiration and eviction doesn't increment it neither, I will open a new issue to discuss it.